File: secureBridge.ts
import * as zmq from "zeromq";
import * as jwt from "jsonwebtoken";
import * as fs from "fs";
import * as msgpack from "msgpack-lite";
import * as protobuf from "protobufjs"; // Para Protobuf
import { CircuitBreaker } from "opossum";
import * as client from "prom-client";
import * as express from "express";
import { createLogger, format, transports } from "winston";

// Winston logger
const logger = createLogger({
  format: format.json(),
  transports: [new transports.Console()],
});

// Métricas Prometheus
const requestCount = new client.Counter({
  name: "requests_total",
  help: "Número total de solicitudes",
  labelNames: ["operation"],
});
const requestLatency = new client.Histogram({
  name: "request_latency_seconds",
  help: "Latencia de solicitudes",
  labelNames: ["operation"],
});

// Servidor Prometheus
const app = express();
app.get("/metrics", async (req, res) => {
  res.set("Content-Type", client.register.contentType);
  res.end(await client.register.metrics());
});
app.listen(8000, () => console.log("Prometheus metrics available on /metrics"));

interface Handler {
  (data: any): any;
}

function BridgeHandler(operation: string) {
  return (target: any, key: string) => {
    if (!target.constructor.handlers) {
      target.constructor.handlers = new Map();
    }
    target.constructor.handlers.set(operation, target[key]);
  };
}

export class SecureBridge {
  private serverSocket: zmq.Reply | null = null;
  private clientSocket: zmq.Request | null = null;
  private handlers: Map<string, Handler> = new Map();
  private jwtSecret: string;
  private circuitBreaker: CircuitBreaker;

  constructor(
    private serverPort: number,
    private clientPort: number,
    private privateKey: string,
    private publicKey: string,
    private peerPublicKey: string,
    jwtSecret: string
  ) {
    this.jwtSecret = jwtSecret;
    this.circuitBreaker = new CircuitBreaker(this.sendRequest.bind(this), {
      errorThresholdPercentage: 50,
      resetTimeout: 30000,
    });
  }

  registerHandler(operation: string, handler: Handler) {
    this.handlers.set(operation, handler);
  }

  validateJwt(token: string): any {
    try {
      return jwt.verify(token, this.jwtSecret);
    } catch (error) {
      throw new Error("Invalid JWT");
    }
  }

  serialize(data: object, format: string): Buffer {
    if (format === "protobuf") {
      const message = BridgeMessage.create(data); // Clase generada por Protobuf
      return BridgeMessage.encode(message).finish();
    }
    if (format === "json") {
      return Buffer.from(JSON.stringify(data));
    } else if (format === "msgpack") {
      return msgpack.encode(data);
    } else {
      throw new Error("Unsupported serialization format");
    }
  }

  deserialize(data: Buffer, format: string): any {
    if (format === "protobuf") {
      const message = BridgeMessage.decode(data); // Clase generada por Protobuf
      return BridgeMessage.toObject(message);
    }
    if (format === "json") {
      return JSON.parse(data.toString());
    } else if (format === "msgpack") {
      return msgpack.decode(data);
    } else {
      throw new Error("Unsupported serialization format");
    }
  }

  @BridgeHandler("sum")
  public handleSum(data: any): any {
    return { result: data.a + data.b };
  }

  async startServer() {
    this.serverSocket = new zmq.Reply();
    this.serverSocket.curve_server = true;
    this.serverSocket.curve_secretkey = this.privateKey;
    this.serverSocket.curve_publickey = this.publicKey;

    await this.serverSocket.bind(`tcp://*:${this.serverPort}`);
    console.log(`Server listening on port ${this.serverPort}`);

    for await (const [msg] of this.serverSocket) {
      const start = Date.now();
      try {
        const message = JSON.parse(msg.toString());
        const token = message.token;
        this.validateJwt(token);
        const operation = message.operation;
        requestCount.inc({ operation });
        const handler = this.handlers.get(operation);
        if (handler) {
          const response = handler(message.data);
          await this.serverSocket.send(
            JSON.stringify({ status: "success", data: response })
          );
        } else {
          throw new Error("Unknown operation");
        }
        requestLatency.observe({ operation }, (Date.now() - start) / 1000);
      } catch (error) {
        logger.error("Server error", { error: error.message });
        await this.serverSocket.send(
          JSON.stringify({ status: "error", message: error.message })
        );
      }
    }
  }

  async sendRequest(message: object, format: string = "json"): Promise<any> {
    if (!this.clientSocket) throw new Error("Client socket not initialized");
    const serializedMessage = this.serialize(message, format);
    await this.clientSocket.send(serializedMessage);
    const [response] = await this.clientSocket.receive();
    return this.deserialize(response, format);
  }

  async startClient(message: object, format: string = "json") {
    this.clientSocket = new zmq.Request();
    this.clientSocket.curve_serverkey = this.peerPublicKey;
    this.clientSocket.curve_secretkey = this.privateKey;
    this.clientSocket.curve_publickey = this.publicKey;

    await this.clientSocket.connect(`tcp://localhost:${this.clientPort}`);
    console.log(`Client connected to port ${this.clientPort}`);

    try {
      return await this.circuitBreaker.fire(message, format);
    } catch (error) {
      console.error("Client error:", error);
      return { status: "error", message: error.message };
    }
  }

  async shutdown() {
    if (this.serverSocket) {
      await this.serverSocket.close();
    }
    if (this.clientSocket) {
      await this.clientSocket.close();
    }
  }
}

File: secure_bridge.py
import zmq
import zmq.auth
from zmq.auth.thread import ThreadAuthenticator
import jwt
import json
import msgpack
import asyncio
import logging
import secrets
import time
import random
from typing import Callable, Dict
from google.protobuf.message import DecodeError  # Para Protobuf
from google.protobuf.json_format import Parse, MessageToJson  # Para Protobuf
from circuitbreaker import CircuitBreaker  # Librería externa para circuit breaker
from fastapi import FastAPI
from prometheus_client import start_http_server, Counter, Histogram
import structlog
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import rsa
import zmq.utils.z85 as z85  # Agregar al inicio con los otros imports
import binascii

# Configuración de logs estructurados
structlog.configure(
    processors=[structlog.processors.JSONRenderer(indent=2)]
)
logger = structlog.get_logger()

app = FastAPI()

REQUEST_COUNT = Counter("requests_total", "Número total de solicitudes", ["operation"])
REQUEST_LATENCY = Histogram("requests_latency_seconds", "Latencia de solicitudes", ["operation"])

@app.get("/openapi.json")
def openapi():
    return app.openapi()

@app.get("/health")
def health_check():
    return {"status": "ok"}

@app.get("/.well-known/jwks.json")
def jwks():
    return {
        "keys": [json.loads(SecureBridge.public_jwk)]
    }

class SecureBridge:
    def __init__(self, server_port: int, client_port: int, private_key: str, public_key: str, peer_public_key: str, jwt_secret: str):
        self.server_port = server_port
        self.client_port = client_port
        # Convertir claves a bytes usando Z85
        try:
            self.private_key = z85.decode(private_key) if isinstance(private_key, str) else private_key
            self.public_key = z85.decode(public_key) if isinstance(public_key, str) else public_key
            self.peer_public_key = z85.decode(peer_public_key) if isinstance(peer_public_key, str) else peer_public_key
        except binascii.Error as e:
            raise ValueError(f"Invalid key format: {e}")
        self.jwt_secret = jwt_secret
        self.context = zmq.Context()
        self.authenticator = None
        self.handlers: Dict[str, Callable] = {}
        self.logger = logging.getLogger("SecureBridge")
        self.logger.setLevel(logging.INFO)
        self.circuit_breaker = CircuitBreaker(
            failure_threshold=3,           # en lugar de fail_max
            recovery_timeout=30,           # en lugar de reset_timeout
            name="secure_bridge_breaker"   # nombre opcional para identificación
        )
        self.last_rotation = time.time()
        self.jwt_rotation_task = None  # Inicialmente None
        self.rsa_private_key = rsa.generate_private_key(
            public_exponent=65537,
            key_size=2048
        )
        self.public_jwk = self.rsa_private_key.public_key().public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo
        )

    @staticmethod
    def generate_curve_keypair():
        """Genera un par de claves CURVE y las devuelve en formato Z85"""
        public, secret = zmq.curve_keypair()
        return z85.encode(secret).decode('ascii'), z85.encode(public).decode('ascii')

    async def init_jwt_rotation(self):
        """Inicializar la tarea de rotación de JWT"""
        self.jwt_rotation_task = asyncio.create_task(self._rotate_jwt_secret())
        return self

    async def _rotate_jwt_secret(self):
        while True:
            await asyncio.sleep(3600)  # Intervalo de rotación de ejemplo
            self.jwt_secret = secrets.token_urlsafe(64)
            self.last_rotation = time.time()
            self.logger.info("JWT secret rotated")

    def start_authenticator(self):
        self.authenticator = ThreadAuthenticator(self.context)
        self.authenticator.start()
        self.authenticator.allow('127.0.0.1')
        self.authenticator.configure_curve(domain='*', location=zmq.auth.CURVE_ALLOW_ANY)

    def stop_authenticator(self):
        if self.authenticator:
            self.authenticator.stop()

    def register_handler(self, operation: str, handler: Callable):
        self.handlers[operation] = handler

    def validate_jwt(self, token: str):
        try:
            return jwt.decode(token, self.jwt_secret, algorithms=["HS256"])
        except jwt.ExpiredSignatureError:
            raise ValueError("JWT expired")
        except jwt.InvalidTokenError:
            raise ValueError("Invalid JWT")

    def serialize(self, data: dict, format: str) -> bytes:
        if format == "protobuf":
            return data.SerializeToString()  # data debe ser un mensaje Protobuf
        elif format == "json":
            return json.dumps(data).encode("utf-8")
        elif format == "msgpack":
            return msgpack.packb(data)
        else:
            raise ValueError("Unsupported serialization format")

    def deserialize(self, data: bytes, format: str) -> dict:
        if format == "protobuf":
            try:
                message = BridgeMessage()  # Clase generada por Protobuf
                message.ParseFromString(data)
                return json.loads(MessageToJson(message))
            except DecodeError as e:
                raise ValueError(f"Protobuf deserialization error: {e}")
        elif format == "json":
            return json.loads(data.decode("utf-8"))
        elif format == "msgpack":
            return msgpack.unpackb(data)
        else:
            raise ValueError("Unsupported serialization format")

    async def start_server(self):
        socket = self.context.socket(zmq.REP)
        socket.curve_secretkey = self.private_key
        socket.curve_publickey = self.public_key
        socket.curve_server = True
        socket.bind(f"tcp://*:{self.server_port}")

        while True:
            try:
                with self.circuit_breaker:
                    message = socket.recv_json()
                    token = message.get("token")
                    self.validate_jwt(token)
                    operation = message.get("operation")
                    data = message.get("data")
                    if operation in self.handlers:
                        response = self.handlers[operation](data)
                        socket.send_json({"status": "success", "data": response})
                        REQUEST_LATENCY.labels(operation=operation).observe(time.time() - start_time)
                    else:
                        socket.send_json({"status": "error", "message": "Unknown operation"})
            except ValueError as e:
                self.logger.error(f"JWT validation error: {e}")
                socket.send_json({"status": "error", "message": str(e)})
            except Exception as e:
                self.logger.error(f"Error in server: {e}")
                socket.send_json({"status": "error", "message": str(e)})

    async def start_client(self, message: dict, format: str = "json"):
        socket = self.context.socket(zmq.REQ)
        socket.curve_secretkey = self.private_key
        socket.curve_publickey = self.public_key
        socket.curve_serverkey = self.peer_public_key
        socket.connect(f"tcp://localhost:{self.client_port}")
        for attempt in range(3):  # Backoff exponencial con jitter
            try:
                serialized_message = self.serialize(message, format)
                socket.send(serialized_message)
                response = socket.recv()
                return self.deserialize(response, format)
            except Exception as e:
                delay = min(2 ** attempt + random.uniform(0, 1), 10)
                self.logger.error(f"Attempt {attempt + 1} failed: {e}, retrying in {delay}s")
                await asyncio.sleep(delay)
        return {"status": "error", "message": "Max retries exceeded"}

    def shutdown(self):
        self.context.term()
        self.stop_authenticator()

    def print_keys(self):
        """Método de diagnóstico para imprimir las claves en un formato legible"""
        print("\n=== Claves CURVE ===")
        print(f"Private Key: {z85.encode(self.private_key).decode('ascii')}")
        print(f"Public Key:  {z85.encode(self.public_key).decode('ascii')}")
        print(f"Peer Public: {z85.encode(self.peer_public_key).decode('ascii')}")
        print("=================\n")